lib/npg_pipeline/function/product_delivery_notifier.pm

```diff
@@ -0,0 +1,335 @@
+
+=cut
+
+sub make_message {
+  my ($self, $product) = @_;
+
+  my $dir_path       = catdir($self->archive_path(), $product->dir_path());
+  my $cram_file      = $product->file_path($dir_path, ext => 'cram');
+  my $cram_md5_file  = $product->file_path($dir_path, ext => 'cram.md5');
+  my $cram_md5       = $self->_read_md5_file($cram_md5_file);
+
+  my $flowcell_barcode = $product->lims->flowcell_barcode();
+  my $sample_id        = $product->lims->sample_id();
+  my $sample_name      = $product->lims->sample_name();
+  my $supplier_name    = $product->lims->sample_supplier_name();
+  $supplier_name or
+    $self->logcroak(sprintf q{Missing supplier name for product %s, %s},
+                    $product->file_name_root(), $product->rpt_list());
+
+  my $message_uuid = uuid();
+  my $subject_uuid = uuid(); # This is a placeholder until we can
+                             # obtain the SScape UUID of the sample
+
+  my $message_body =
+    {
+     'lims'  => $EVENT_LIMS_ID,
+     'event' =>
+     {'event_type'      => $EVENT_TYPE,
+      'uuid'            => $message_uuid,
+      'occurred_at'     => undef,
+      'user_identifier' => $EVENT_USER_ID,
+      'subjects'        =>[{
+                            'role_type'     => $EVENT_ROLE_TYPE,
+                            'subject_type'  => $EVENT_SUBJECT_TYPE,
+                            'friendly_name' => $supplier_name,
+                            'uuid'          => $subject_uuid,
+                           }],
+      'metadata' =>
+      {
+       'customer_name'        => $self->customer_name,
+       'sample_id'            => $sample_id,
+       'sample_name'          => $sample_name,
+       'sample_supplier_name' => $supplier_name,
+       'flowcell_barcode'     => $flowcell_barcode,
+       'file_path'            => $cram_file,
+       'file_md5'             => $cram_md5,
+       'rpt_list'             => $product->rpt_list,
+      }
+     }
+    };
+
+  return $message_body;
+}
+
+=head2 create
+
+  Arg [1]    : None
+
+  Example    : my $defs = $obj->create
+  Description: Create per-product function definitions.
+
+  Returntype : ArrayRef[npg_pipeline::function::definition]
+
+=cut
+
+sub create {
+  my ($self) = @_;
+
+  if (! -d $self->message_dir) {
+    make_path($self->message_dir);
+  }
+
+  my $id_run = $self->id_run();
+  my @definitions;
+
+  my $i = 0;
+  foreach my $product (@{$self->products->{data_products}}) {
+    next if not $self->is_release_data($product);
+    next if not $self->has_qc_for_release($product);
+    next if not $self->is_for_s3_release($product);
+    next if not $self->is_for_s3_release_notification($product);
+
+    my $msg_file = $product->file_path($self->message_dir, ext => 'msg.json');
+    $self->_write_message_file($msg_file, $self->make_message($product));
+
+    my $job_name = sprintf q{%s_%d_%d}, $SEND_MESSAGE_SCRIPT, $id_run, $i;
+    my $command =
+      sprintf q{%s --host %s --port %d --vhost %s --exchange %s},
```

Comment:
Missing commit? Do these options need to be replaced with a `-c ~/.npg/psd_production_events.conf`?